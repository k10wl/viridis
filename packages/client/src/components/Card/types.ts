export type CardT = {
  views: number;
  favorite: boolean;
  imageSrc: string;
  name: string;
  description: string;
  tags: string[];
  onClick: () => void;
};
