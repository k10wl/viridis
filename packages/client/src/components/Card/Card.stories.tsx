import { ComponentMeta, ComponentStory } from "@storybook/react";
import React from "react";

import { Card } from "./index";

export default {
  title: "Card",
  component: Card,
} as ComponentMeta<typeof Card>;

const Template: ComponentStory<typeof Card> = (props) => <Card {...props} />;

export const ThreeIngredients = Template.bind({});

ThreeIngredients.args = {
  views: 0,
  favorite: false,
  imageSrc:
    "https://welcome.setantasports.com/wp-content/uploads/2022/03/placeholder.png",
  name: "Name of card",
  description: "Recipe description",
  ingredients: ["Milk", "Cocoa", "Sugar"],
  onClick: (): void => undefined,
};

export const IngredientsOveflow = Template.bind({});

IngredientsOveflow.args = {
  views: 0,
  favorite: false,
  imageSrc:
    "https://welcome.setantasports.com/wp-content/uploads/2022/03/placeholder.png",
  name: "Name of card",
  description: "Recipe description",
  ingredients: [
    "Egg",
    "Bread",
    "Sugar",
    "Flour",
    "Rehydrated water",
    "Bacon",
    "Poultry",
    "Turkey legs",
    "Chicken breast",
    "Mango",
    "Cheese",
    "Mozzarella",
    "Gorgonzola",
    "Some rare blue cheese name",
    "Soy beans",
    "Honey",
    "Mush",
    "Bananas",
    "Garlic",
    "Onion",
    "Milk",
    "Cocoa",
    "Salt",
  ],
  onClick: (): void => undefined,
};

export const LongDescriptionAndName = Template.bind({});

LongDescriptionAndName.args = {
  views: 0,
  favorite: false,
  imageSrc:
    "https://welcome.setantasports.com/wp-content/uploads/2022/03/placeholder.png",
  name: "Recipe name recipe name recipe name recipe name recipe name recipe name recipe name recipe name recipe name",
  description:
    "Recipe description recipe description recipe description recipe description recipe description recipe description recipe description recipe description recipe description recipe description",
  ingredients: ["Egg", "Bread", "Sugar", "Flour", "Chicken breast"],
  onClick: (): void => undefined,
};
