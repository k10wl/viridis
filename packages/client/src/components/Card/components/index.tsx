export * from "./CardHeader";
export * from "./CardContent";
export * from "./TagsContainer";
