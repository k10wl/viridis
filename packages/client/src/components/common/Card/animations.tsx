import { Variants, Transition } from "framer-motion";

import {
  accentColor2,
  backgroundColor1,
  backgroundColor3,
} from "../../../stylesheets/Colors";

export const cardBaseVariants: Variants = {
  rest: {
    backgroundColor: backgroundColor1,
  },
  hover: {
    backgroundColor: backgroundColor3,
    boxShadow: `0 0 5px 0 ${accentColor2}`,
  },
  tap: {
    backgroundColor: backgroundColor3,
    boxShadow: `0 0 0px 0 ${accentColor2}`,
  },
};

export const cardBaseTransition: Transition = {
  ease: "linear",
};
