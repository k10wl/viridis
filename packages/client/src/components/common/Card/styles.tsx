import styled from "styled-components";
import { motion } from "framer-motion";

import { accentColor2, accentColor4 } from "src/stylesheets/Colors";

export const CardBase = styled(motion.div).attrs({
  tabIndex: 0,
})`
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  cursor: pointer;

  overflow: hidden;

  width: 340px;
  height: 550px;

  border: 2px solid ${accentColor2};
  border-radius: 7px;

  margin: 20px;

  &:focus-visible {
    outline: 3px solid ${accentColor4}80;
    outline-offset: 3px;
  }
`;
