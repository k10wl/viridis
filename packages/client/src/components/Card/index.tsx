import React, { useCallback } from "react";

import { CardContent, CardHeader, TagContainer } from "./components";
import { CardBase } from "./styles";
import { cardBaseTransition, cardBaseVariants } from "./animations";
import { CardT } from "./types";

const Card: React.FC<CardT> = ({
  views,
  favorite,
  description,
  imageSrc,
  name,
  tags,
  onClick,
}) => {
  const onEnterClick = useCallback(
    (keyboardEvent: React.KeyboardEvent<HTMLDivElement>) => {
      if (keyboardEvent.key === "Enter") {
        onClick();
      }
    },
    []
  );

  return (
    <CardBase
      variants={cardBaseVariants}
      initial="rest"
      whileHover="hover"
      whileTap="tap"
      transition={cardBaseTransition}
      onClick={onClick}
      onKeyPress={onEnterClick}
      data-testid="card"
    >
      <div>
        <CardHeader views={views} favorite={favorite} />
        <CardContent
          imageSrc={imageSrc}
          name={name}
          description={description}
        />
      </div>
      <TagContainer tags={tags} />
    </CardBase>
  );
};

export default Card;
