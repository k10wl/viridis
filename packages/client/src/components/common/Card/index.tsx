import React, { useCallback } from "react";

import { CardContent, CardHeader, TagsContainer } from "./components";
import { CardBase } from "./styles";
import { cardBaseVariants, cardBaseTransition } from "./animations";

export type CardT = {
  views: number;
  favorite: boolean;
  imageSrc: string;
  name: string;
  description: string;
  tags: string[];
  onClick: () => void;
};

const Card: React.FC<CardT> = ({
  views,
  favorite,
  description,
  imageSrc,
  name,
  tags,
  onClick,
}) => {
  const onEnterClick = useCallback(
    (keyboardEvent: React.KeyboardEvent<HTMLDivElement>) => {
      if (keyboardEvent.key === "Enter") {
        onClick();
      }
    },
    []
  );

  return (
    <CardBase
      variants={cardBaseVariants}
      initial="rest"
      whileHover="hover"
      whileTap="tap"
      transition={cardBaseTransition}
      onClick={onClick}
      onKeyPress={onEnterClick}
      data-testid="card"
    >
      <div>
        <CardHeader views={views} favorite={favorite} />
        <CardContent
          imageSrc={imageSrc}
          name={name}
          description={description}
        />
      </div>
      <TagsContainer tags={tags} />
    </CardBase>
  );
};
export default Card;
