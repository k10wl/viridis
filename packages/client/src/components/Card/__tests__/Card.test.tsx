import React from "react";
import reactRenderer from "react-test-renderer";

import { fireEvent, render, screen } from "@testing-library/react";

import Card from "../index";

const properties = {
  views: 3,
  favorite: false,
  imageSrc: "Eheu, homo!",
  name: "Orexis, orgia, et ausus.",
  description: "Cur turpis observare?",
  tags: ["Planeta", "Brodiums"],
  onClick: jest.fn(),
};

test("Card component", () => {
  const tree = reactRenderer
    .create(
      <Card
        views={properties.views}
        favorite={properties.favorite}
        imageSrc={properties.imageSrc}
        name={properties.name}
        description={properties.description}
        tags={properties.tags}
        onClick={properties.onClick}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();

  render(
    <Card
      views={properties.views}
      favorite={properties.favorite}
      imageSrc={properties.imageSrc}
      name={properties.name}
      description={properties.description}
      tags={properties.tags}
      onClick={properties.onClick}
    />
  );

  const cardNode = screen.getByTestId("card");
  const nameNode = screen.getByText(properties.name);
  const viewsNode = screen.getByText(properties.views);
  const imageNode = screen.getByTestId("image");
  const favoriteNode = screen.getByTestId("favorite-button");
  const descriptionNode = screen.getByText(properties.description);

  expect(cardNode).toBeInTheDocument();
  expect(nameNode).toBeInTheDocument();
  expect(viewsNode).toBeInTheDocument();
  expect(imageNode).toBeInTheDocument();
  expect(imageNode).toHaveAttribute("src", properties.imageSrc);
  expect(favoriteNode).toBeInTheDocument();
  expect(descriptionNode).toBeInTheDocument();

  properties.tags.forEach((tag) => {
    const tagNode = screen.getByText(tag);

    expect(tagNode).toBeInTheDocument();
  });

  fireEvent.click(cardNode);
  expect(properties.onClick).toHaveBeenCalledTimes(1);
  properties.onClick.mockReset();

  fireEvent.keyPress(cardNode, { key: "Enter", code: "Enter", charCode: 13 });
  expect(properties.onClick).toHaveBeenCalledTimes(1);
});
