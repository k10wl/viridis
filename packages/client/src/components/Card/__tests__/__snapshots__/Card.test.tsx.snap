// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  padding: 10px 10px 5px;
}

.c2 {
  padding: 0 5px;
  color: #696262;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 18px;
  height: 18px;
  -webkit-transform: translateY(-2px);
  -ms-transform: translateY(-2px);
  transform: translateY(-2px);
  border: none;
  border-radius: 50%;
  padding: 3px 0 0;
  margin: 0;
  background-color: transparent;
  cursor: pointer;
}

.c3:focus-visible {
  outline: 3px solid #c6725880;
  outline-offset: 3px;
}

.c4 {
  height: 280px;
  width: 100%;
  object-fit: cover;
}

.c5 {
  padding: 10px 10px 10px;
  color: #121111;
  font-family: "PT Serif Bold",serif;
  font-size: 1.125rem;
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c6 {
  padding: 0 10px;
  color: #696262;
  font-family: "Roboto Thin",sans-serif;
  font-size: 0.875rem;
  text-overflow: ellipsis;
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: 3;
  -webkit-box-orient: vertical;
  white-space: normal;
}

.c10 {
  min-width: 150px;
  height: 50px;
  outline: none;
  border: none;
  border-radius: 4px;
  font-family: "Roboto",sans-serif;
  font-size: 1rem;
}

.c10:hover {
  cursor: pointer;
}

.c10:focus-visible {
  outline: 3px solid #c6725880;
  outline-offset: 1px;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  height: 143px;
  margin-top: auto;
  padding: 10px 10px;
  background-color: #f2f3f5;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 5px;
  max-height: 66px;
  overflow: hidden;
  list-style-type: none;
}

.c9 {
  padding: 3px 10px;
  border-radius: 4px;
  text-align: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  color: #fbfdfd;
  background-color: #244e4b;
  font-family: "Roboto Thin",sans-serif;
  font-size: 0.75rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  cursor: pointer;
  overflow: hidden;
  width: 340px;
  height: 550px;
  border: 2px solid #244e4b;
  border-radius: 7px;
  margin: 20px;
}

.c0:focus-visible {
  outline: 3px solid #c6725880;
  outline-offset: 3px;
}

<div
  className="c0"
  data-testid="card"
  onClick={[MockFunction]}
  onKeyPress={[Function]}
  style={
    Object {
      "backgroundColor": "#fbfdfd",
    }
  }
  tabIndex={0}
>
  <div>
    <div
      className="c1"
    >
      <span>
        <svg
          fill="#bebab9"
          width={14}
        >
          eye-regular.svg
        </svg>
        <span
          className="c2"
        >
          3
        </span>
      </span>
      <button
        className="c3"
        data-testid="favorite-button"
        onClick={[Function]}
      >
        <svg
          data-testid="regular-icon"
          fill="#bebab9"
          style={
            Object {
              "opacity": 1,
              "transform": "none",
              "transformOrigin": "center",
            }
          }
          width="100%"
        >
          heart-regular.svg
        </svg>
      </button>
    </div>
    <img
      alt="Orexis, orgia, et ausus."
      className="c4"
      data-testid="image"
      src="Eheu, homo!"
    />
    <p
      className="c5"
    >
      Orexis, orgia, et ausus.
    </p>
    <p
      className="c6"
    >
      Cur turpis observare?
    </p>
  </div>
  <div
    className="c7"
  >
    <ul
      className="c8"
    >
      <li
        className="c9"
      >
        Planeta
      </li>
      <li
        className="c9"
      >
        Brodiums
      </li>
    </ul>
    <button
      className="c10"
      disabled={false}
      height={50}
      onClick={[Function]}
      style={
        Object {
          "backgroundColor": "#c67258",
          "border": "#c67258 solid 0",
          "color": "#fbfdfd",
        }
      }
      type="button"
    >
      Open recipe in new page
    </button>
  </div>
</div>
`;
